File: app/src/main/java/com/gatheringhallstudios/mhworlddatabase/common/adapters/MonsterListAdapter.java
package com.gatheringhallstudios.mhworlddatabase.common.adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.gatheringhallstudios.mhworlddatabase.R;
import com.gatheringhallstudios.mhworlddatabase.data.views.Monster;

import java.util.List;

/**
 * Adapter to display a list of Monsters
 */

public class MonsterListAdapter extends RecyclerView.Adapter<MonsterListAdapter.ViewHolder> {

    private List<Monster> monsterList;

    public MonsterListAdapter(List<Monster> monsters) {
        monsterList = monsters;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater lf = LayoutInflater.from(parent.getContext());
        View view = lf.inflate(R.layout.listitem_monster, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Monster monster = monsterList.get(position);

        holder.monsterName.setText(monster.name);
        // TODO Set monster image
    }

    @Override
    public int getItemCount() {
        return monsterList.size();
    }

    public void replaceData(List<Monster> list) {
        monsterList = list;
        notifyDataSetChanged();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        ImageView monsterIcon;
        TextView monsterName;

        ViewHolder(View itemView) {
            super(itemView);
            monsterIcon = itemView.findViewById(R.id.monster_icon);
            monsterName = itemView.findViewById(R.id.monster_name);
        }
    }
}

File: app/src/main/java/com/gatheringhallstudios/mhworlddatabase/features/monsters/list/MonsterListAdapter.kt
package com.gatheringhallstudios.mhworlddatabase.features.monsters.list

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.gatheringhallstudios.mhworlddatabase.R
import com.gatheringhallstudios.mhworlddatabase.adapters.common.SimpleRecyclerViewAdapter
import com.gatheringhallstudios.mhworlddatabase.adapters.common.SimpleViewHolder
import com.gatheringhallstudios.mhworlddatabase.assets.AssetLoader
import com.gatheringhallstudios.mhworlddatabase.data.models.MonsterBase
import com.gatheringhallstudios.mhworlddatabase.getRouter
import com.gatheringhallstudios.mhworlddatabase.databinding.ListitemMonsterBinding


/**
 * A RecyclerView adapter for a homogeneous monster list
 */
class MonsterListAdapter: SimpleRecyclerViewAdapter<MonsterBase>() {
    override fun onCreateView(parent: ViewGroup): View {
        val inflater = LayoutInflater.from(parent.context)
        return inflater.inflate(R.layout.listitem_monster, parent, false)
    }

    override fun bindView(viewHolder: SimpleViewHolder, data: MonsterBase) {
        val icon = AssetLoader.loadIconFor(data)
        val binding = ListitemMonsterBinding.bind(viewHolder.itemView)
        binding.icon.setImageDrawable(icon)
        binding.monsterName.text = data.name

        if (data.ecology.isNullOrEmpty()) {
            binding.ecologyText.visibility = View.GONE
        } else {
            binding.ecologyText.visibility = View.VISIBLE
            binding.ecologyText.text = data.ecology
        }

        viewHolder.itemView.setOnClickListener { it.getRouter().navigateMonsterDetail(data.id) }
    }
}
